## !!! To be updated !!! 
## Fine-mapping of Prostate Cancer Susceptibility Loci in a Large Meta-Analysis Identifies Candidate Causal Variants. 

#### Tokhir Dadaev<sup>1</sup>, Edward J. Saunders<sup>1</sup>, etc ... , Eeles RA<sup>2</sup>, Kote-Jarai Z<sup>45</sup>, *et al.*


### **Abstract**

We used a novel multivariate fine-mapping strategy to 84 published prostate cancer GWAS loci that replicated in a meta-analysis of 82,591 cases and 61,213 controls of European ancestry. We applied the Bayesian variable selection algorithm JAM, which explores multi-SNP models in summary data while accounting for linkage disequilibrium, to infer the number of independent signals and the credible set of SNPs driving those associations.

For 63 regions we observed a single signal and in 12 regions we show evidence for multiple independent signals, up to a maximum of 5, bringing the total number of hits across all regions to 99. Furthermore, only 15 original GWAS tag SNPs remained within the catalogue of candidate variants, with all other signals being replaced by more likely candidates.

Biological annotation of all variants indicated significant enrichment within the credible set for variants in promoter and enhancer elements, intersecting binding sites for several transcription factors including AR, ERG and FOXA1, within transcribed regions including variants altering protein structure, and in conserved segments of the genome. Additionally, in 40 regions at least one variant was co-localised with an eQTL in prostate cancer tissue.

The new, more accurate set of 99 candidate variants identified substantially increased the proportion of familial relative risk explained by these known GWAS regions from 23.5% to 31.8%. This highlights the importance of fine-mapping studies in accounting for some of the 'missing heritability' of PrCa and has important implications for clinical risk profiling.


[PMID: XXX](http://www.ncbi.nlm.nih.gov/pubmed/XXX)

