## !!! To be updated !!! 
## Fine-mapping of Prostate Cancer Susceptibility Loci in a Large Meta-Analysis Identifies Candidate Causal Variants. 

#### Dadaev T<sup>2</sup>, Saunders EJ<sup>2</sup>, etc ... , Eeles RA<sup>2</sup>, Kote-Jarai Z<sup>45</sup>, *et al.*


### **Abstract**

We fine-mapped 88 published prostate cancer GWAS loci that replicated in a meta-analysis of 82,591 cases and 61,213 controls of European ancestry. We used Bayesian variable selection to determine the number of independent signals and the credible set of SNPs driving those signals. Specifically, the multivariate fine-mapping algorithm JAM, which explores multi-SNP models in summary data while accounting for linkage disequilibrium, was applied to each region. For 65 regions we observed a single risk signal, with a median 13 candidate SNPs per region driving those associations. At 16 loci, we show evidence for multiple independent prostate cancer risk signals, with a median 41.5 candidate variants per region and a maximum of 5 signals. Biological annotation indicated significant enrichment for variants in promoter and enhancer elements, within transcribed regions, including variants altering protein structure, and in conserved segments of the genome. Additionally, in 39 regions at least one variant was co-localised with an eQTL in prostate cancer tissue. The new, more accurate set of candidate variants and novel association signals identified substantially increased the proportion of familial relative risk explained by these known GWAS regions from 21.7 to 27.4%. 

[PMID: XXX](http://www.ncbi.nlm.nih.gov/pubmed/XXX)

